meetings/working-groups/collection-literals/CL-2023-08-03.md

```diff
@@ -0,0 +1,140 @@
+# Collection literals working group meeting for August 3, 2023
+
+## Agenda
+
+* Collection expressions and ref-structs (like `ReadOnlySpan<T>`).
+* Confirm behavior collection expressions and `ReadOnlySpan`s of blittable data.
+
+## Discussion
+
+The purpose of the meeting was to discuss what semantics we would have in the language for a collection literal target-typed to a ref-struct.
+
+First, for simplicity and to aid by giving examples, we'll use `ReadOnlySpan<T>` to demonstrate the general problem space, the varying options, and the pros/cons of each option.  Second, all these cases deal with collection expression elements of *non-blittable-constant data* (e.g. `[a, b, c]` vs `[1, 2, 3]`).  Rules around the latter are covered at the end.
+
+
+Today, a user can write code like so:
+
+```c#
+ReadOnlySpan<int> x1 = stackalloc int[] { a, b, c }; // stack allocated, local scope.
+ReadOnlySpan<int> x2 = new int[] { a, b, c };        // heap allocated, global scope.
+```
+
+Based on how the variable is initialized, the language then dictates rules about how it can be used.  For example, with the above code it would be illegal to `return x1;` from the method (since it has local scope), while it would be fine to `return x2;`.  Variables without an initializer are assumed to have global scope.  So the following, for example, would be illegal:
+
+```c#
+ReadOnlySpan<int> x3;              // global scope
+x3 = stackalloc int[] { a, b, c }; // illegal.  global scope now points at local data.
+```
+
+Unlike existing constructs, which clearly delineate whether they are on the stack or heap, a decision needs to be made as to what the following means:
+
+```c#
+ReadOnlySpan<int> x4 = [a, b, c];
+```
+
+The working group considered several options.
+
+### Option 1.  Stack-alloc'ed only if target has local-scope.
+
+The first option considered was that a collection expression used with a `ReadOnlySpan<T>` would itself only be stack-alloc'ed if the span itself had an explicitly stated local scope.   In other words, the following:
```

Comment:
The target may have an implicit scope that we'd use too.
```c#
ReadOnlySpan<int> x3 = stackalloc int[1];  // x3 has local scope
...
x3 = [a, b, c]; // stack allocated, local scope
```

Reply:
fair.  let me try to revise.